Initial PPP implementation

jukkar opened the pull request:
Initial implementation for Point-to-point protocol (PPP). It implements LCP, IPCP and IPV6CP modules. See these RFCs for details.
* LCP : https://tools.ietf.org/html/rfc1661
* HDLC : https://tools.ietf.org/html/rfc1662
* IPCP : https://tools.ietf.org/html/rfc1332
* IPV6CP : https://tools.ietf.org/html/rfc5072

This is still work-in-progress and following things need more TLC:
- [ ] Authentication support missing
- [ ] LCP non-default option negotiation
- [ ] MRU/MTU handling not yet done

Fixes #14034

jukkar committed:
net: ppp: Initial support for point-to-point protocol

This implements ppp L2 component, LCP and IPCP modules.

Fixes #14034

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
drivers: net: ppp: Driver for point-to-point protocol

The ppp driver uses uart_pipe driver for data transfer.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
cmake: Add serial port for ppp in qemu

In order to simplify ppp testing, use ppp specific serial port
when starting qemu.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: pkt: Add net_pkt_write_le16() helper

We had a big endian helper but little endian one was missing.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
tests: net: ppp: Add unit tests for PPP driver

Make sure that we are able to parse incoming PPP data
correctly.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: if: No need to set IP addresses etc for PPP

If the network interface is point-to-point one which does
not need IP address etc, then no need to start DAD etc for
those interfaces.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: pkt: Add net_pkt_read_le16() helper

We had a big endian helper but little endian one was missing.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: ppp: Allow delay of PPP protocol handshakes

By default PPP is started immediately when the network interface
goes up. This can be problematic especially when debugging the beast
so allow user to delay the startup.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: ppp: Add IPV6CP support

Initial version for PPP IPv6 Control Protocol.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: ppp: Reject unhandled protocols

If we receive a protocol that we do not currently handle, then
return Protocol-Reject to peer.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Resolved merge conflicts

jukkar committed:
samples: net: echo-server: Add PPP overlay config file

A sample overlay config file for PPP added.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
First version that works properly against Linux pppd with IPv4 and IPv6.

jukkar committed:
net: ppp: Add proper support to receive Echo-Reply message

Currently only net-shell calls net_ppp_ping() command, so make
it return the amount of time that it took to receive Echo-Reply
so the net-shell can print the round trip time value.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
* rebased against latest master
* added support for sending Echo-Request and receiving Echo-Reply from net-shell

jukkar committed:
net: ppp: Handle received Discard-Request

We must discard the received Discard-Request silently.
See RFC 1661 chapter 5.9 for details.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
doc: net: ppp: Add PPP documentation

Add information about PPP into Zephyr documentation.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

tbursztyka reviewed (changes requested):
Most of it seems fine to me. 

There is just (well it is not a small change) the logic of config_info_* functions and ppp_send_pkt: that has to be reworked so net_pkt is allocated first so all its API can be used instead of the net_buf one.
Whit that done, I think it will be good to go.

jukkar commented on subsys/net/l2/ppp/ppp_stats.c:
```diff
@@ -0,0 +1,55 @@
+	const struct ppp_api *ppp;
+
+	switch (NET_MGMT_GET_COMMAND(mgmt_request)) {
+	case NET_REQUEST_STATS_CMD_GET_PPP:
```
Yes, overkill, will change.

jukkar commented on subsys/net/l2/ppp/fsm.c:
```diff
@@ -0,0 +1,1088 @@
+		 enum ppp_packet_type type, u8_t id,
+		 void *data, u32_t data_len)
+{
+	struct net_pkt *req_pkt = data;
```
Yes, for PROTOCOL_REJECT and CODE_REJECT packets we re-use existing variable. I didn't want to introduce new variable into the function. I will add some more comments here.

jukkar commented on subsys/net/l2/ppp/fsm.c:
```diff
@@ -0,0 +1,1088 @@
+		   type == PPP_CONFIGURE_REJ || type == PPP_CONFIGURE_NACK) {
+		/* add options */
+		if (data) {
+			net_buf_frag_add(pkt->frags, data);
```
yes, it contains the options we want to send.

jukkar commented on subsys/net/l2/ppp/fsm.c:
```diff
@@ -0,0 +1,1088 @@
+		return NET_DROP;
+	}
+
+	ret = net_pkt_read_be16(pkt, &length);
```
Yes, it could be done like that too. The PPP protocol does not specify that kind of struct in the RFC so I did it by reading the values individually. I do not want to start to change that at this point, but can be tweaked later.

jukkar commented on include/net/ppp.h:
```diff
@@ -0,0 +1,539 @@
+				       u16_t length);
+
+		/** Add Configuration Information */
+		struct net_buf *(*config_info_add)(struct ppp_fsm *fsm);
```
Yes, I agree that this is a bit cumbersome. The problem was that I needed to construct the options before I had net_pkt, thus only net_buf was used. This is in the todo list and will be handled but a bit later as this is bigger change.

jukkar commented on subsys/net/l2/ppp/Kconfig:
```diff
@@ -11,6 +11,16 @@ menuconfig NET_L2_PPP
+	help
+	  If the PPP starts too fast, it is possible to delay it
+	  a bit. This is mostly useful in debugging if you want the
+	  device be fully up before PPP handshake is started.
```
I used the option for testing purposes. Normally the ppp comes up so fast that lot of logging is lost which was very annoying. So created this config option in order to help debugging things. Also enabled this option in echo-server sample because of that. Default value is 0 which means no delays.

jukkar commented on include/net/ppp.h:
```diff
@@ -349,6 +349,9 @@ struct ipv6cp_options {
 /** PPP L2 context specific to certain network interface */
 struct ppp_context {
+	/** PPP startup worker. */
+	struct k_delayed_work startup;
```
I am not really expecting there to be more than one context which means that there would be only one ppp network interface in the system. This is not the same thing we had with Ethernet and VLANs. So if this becomes really a problem, then we start to think about optimizing this. I will add a todo comment about this in the code.

jukkar replied:
Additional note about this. The PPP driver uses uart_pipe driver which supports only one serial connection, so actually we cannot have more than one PPP network interface currently.

jukkar commented:
> There is just (well it is not a small change) the logic of config_info_* functions and ppp_send_pkt: that has to be reworked so net_pkt is allocated first so all its API can be used instead of the net_buf one.
> Whit that done, I think it will be good to go.

This is in the TODO list (also marked in the code), but as this is a bigger change, I want to postpone this a bit and to be done as a bug fix.

jukkar commented:
Updated according to comments.

jukkar commented:
@dbkinder I added a commit "doc: net: ppp: Add PPP documentation " after you had reviewed the code. Perhaps you want to check it too.

jukkar commented:
Fixed merge conflict.

SebastianBoe reviewed (approved):
Build and configuration related changes are OK

jukkar commented:
Fixed merge conflict

jukkar committed:
net: shell: Refactor network statistics printing

The traffic class printing section was too convoluted because of
many #ifdef's.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: shell: Add ppp network interface support

Print point-to-point network information properly.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
boards: Set support for serial port networking as netif:serial-net

As we now have PPP support, use more generic "serial-net" string instead
of "slip" when setting what kind of networking the board supports.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Fixed the sanitychecker errors.

jukkar commented:
Resolved merge conflicts.

tbursztyka reviewed (approved):
Let's get this in. However, the logic of creation packet with options will have to be reworked, too late for this now.

jukkar commented on drivers/net/ppp.c:
```diff
@@ -0,0 +1,633 @@
+		LOG_DBG("Saving byte %02x", byte);
+	}
+
+	/* This is not very intuitive but we must allocate new buffer
```
Hmm, I prefer leaving this comment here for time being. I spent hours debugging this as it was not clear why the api requires that. At least for me, it would be difficult to understand why we have the ppp->available==1 check without the comment. The documentation of net-pkt should be improved in this respect.
